### ISSUE TITLE
do we need cors mode?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-client#4]
I am not sure whether it is possible to turn off security in every major browser. If not, then we need some sort of workaround at least in the example codes. I am not sure whether the client can do anything about it, or we just need to write a working example with CORS. I need to check this in different browsers. IE will be a pain in the ass I guess.

**Comment by inf3rno** [comment]
I would stick to disabled browser security for now, because that makes things simpler and we can test 3rd party web applications or even write client bots with it. I just need to check which browsers support disabled security and accessing the frame without CORS. There is another header, which prevents putting a page into an iframe. I don't remember which one that is, but I need to check whether disabled security turns that check off too. This is related to #3 too.

**Comment by inf3rno** [comment]
I would make this a sub-issue of #6 and the other one too.

**Comment by inf3rno** [comment]
There are other security headers too. For example security policy, etc. I don't really remember, which ones are important here, but I read about them a few years ago. I need to check what effect they have with disabled and enabled browser security. I think supporting a CORS mode is not a priority for now, but later it may be important to tests client side security features of the page and to write bots for potentially unsafe pages. For stable navigation it is enough now to test the other browsers whether they support disabled security and iframes properly, but I can continue this issue after that with lower priority to add a CORS mode. I might need a different karma config or CLI configuration for that.

**Comment by inf3rno** [comment]
We definitely need a CORS mode, so we don't have to disable browser security. I think the tested page should have a testing mode, which allows the test script to put it into an iframe. I'll try out putting it into a window too, but I am not sure it will be any different.

**Comment by inf3rno** [comment]
We definitely need a CORS mode and a different client for non-Chrome browsers. https://blog.carbonfive.com/2012/08/17/cross-domain-browser-window-messaging-with-html5-and-javascript/

Even if we give the CORS headers, the cross-frame protection of Firefox will kick in to prevent clickjacking or any manipulation of the opened or opener window. This really sucks, because the only way to circumvent this to use `window.postMessage` or any other cross-window communication solution, cookies etc...

The real shit comes only after this, because the receiver window does not even get info about the sender window, just the origin, so we need to implement some kind of handshake protocol, etc. to separate messages coming from different childs or going to just a single child by doing tests in parallel. So this is a lot more than just adding a CORS header so I think this must be delayed and concentrate only on Chrome for now. Maybe this will change later in Chrome too, which will have severe consequences on this lib. But for now it is safe...

As far as I remember there was a technique maybe 10-15 years ago to send js code to the client and eval it. I think the parent window should use that technique to gain full control over the child window instead of translating every message to an operation or operation group.

For example a problematic code in the client now is the following:

```js
var timer = setInterval(function () {
	if (oldURI == this.win.location.href)
		return;
	clearInterval(timer);
	// the URI is set on the child window, we can add the 'load' event handler
}.bind(this), 1);
```

We could do something like:

```js
var childLocation = await this.win.execute(function (childWindow, next){
	var timer = setInterval(function () {
		if (oldURI == childWindow.location.href)
			return;
		clearInterval(timer);
		next(childWindow.location.href);
	}.bind(this), 1);
});
```

Ofc. this is a very bad example, because the communication script can be loaded only after the URI is set on the child window, but I did not want to write a jquery example just for this, so let's imagine that that script is running somehow in the background and it can capture when the URI is set.

For the communication we need to call `func.toString()` on the function, send it to the child window along with a windowId and a requestId. In the child window we need to eval the string and compile the function. After that we call it with the local `window` object and the `next` function. The `next` function should get the results and respond to the parent with the given identifiers, so the parent window will know which `await` it should continue.

Pairing the child windows with the parent windows is not trivial, we could do a queue and wait for each window to load, so by multiple parents we will know for certain which child belongs to which parent. The funny part that because of the design of postMessage, this is not necessary. I mean if there is a window opened with the right URI, and we cannot access it directly, just via postMessage, then nobody cares about who opened that window, we just care about what URI it loaded. If the right URI is loaded for 2 parents, then both will get 1-1 child window, they can communicate with via postMessage. The problematic part that the comm is reloaded in the child window, so it won't remember its id. If we are testing a something, which is dependent on the history of the actual page e.g. we POSTed a form to get to the URI, not just loaded it via GET. Then this can be a problem. By writing this I had the brilliant idea to have an e2e-server on the actual domain, which opens the child windows or maybe even iframes and does the controlling. So only the x-frame-options could affect it, but the CORS not. We could even open it in a child window, and if so we can make it certain that only 1 e2e-server is loaded, which can serve multiple e2e clients at once. 1 server for one domain is enough because it communicates with the client via postMessage, not via the classic parent-child js. So this one e2e-server is opened by one of the e2e-clients testing the same application on the same domain, and after it is loaded, it can server all of these clients alone. Duplication of servers would cause doing the tests twice, so we need to protect against that, for example by sending a heartbeat with the e2e-server to the clients, and listen to the heartbeat before opening a new server. Another possible solution that a new server should send out a message looking for other instances on the same domain. In both cases when the server is lost because of some reason, the clients can open a new one. In the first case the heartbeat is lost with the server. In the second case the messages sent by the clients will timeout, so they need to resend them after checking that the server is not responding. Both solutions have their own issues. I might implement both of them I am not sure yet. The e2e server will open the requested pages in child windows or in iframes if x-frame-options is set to SAMEORIGIN instead of DENY. This way it can make a childWindow - sessionId map, which it can use to communicate with the e2e clients, and so the id-s of the child windows won't lost by page reload.

So for doing this with CORS we need a page, which is served on the tested domain and which serves the e2e server. This can even run on the production server to measure response times from different locations if we protect it enough. The other option to run it in debug mode only. All it should do is loading a js file, which we give and set CORS headers for the domain the e2e client is running on. So something like:

```js
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'OPTIONS, POST, GET, PUT, DELETE');
  res.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
```

```html
<script src="server.js"></script>
<script>
var server = new e2e.Server({
    communication: e2e.communication-post-message,
    container: e2e.container-iframe,
    heartBeat: 100,
    timeout: 10000
});
server.run();
</script>
```

or it can be just 

```html
<script src="server.js"></script>
<script src="run-default-server.js"></script>
```

We can protect it by putting it outside of www root and removing the htaccess which directs to the e2e server in the deploy process. Another option to check the environment and choose not to route to the testing pages if it is production. It would be better to have a dev directory for this, but it doesn't really matter. It would be fun to run the test scripts on the production server as part of the deploy process, so we can be sure that the code works properly on the production server too.

In our tests we can do:

```html
<script src="client.js"></script>
<script>
var client = new e2e.Client({
    communication: e2e.communication-post-message,
    heartBeat: 100,
    timeout: 10000,
    server: "http://localhost:3333/testing/e2e/index.html",
    baseURI: "http://localhost:3333"
});

await client.load("/");
...
</script>
```

We can send the whole test to the server, and use it as a test server, or just the commands we need it to do. I'll decide about that later.

I think the discussion about this is done, and we can start a new issue for the implementation. This should be after the API is working with the disabled security direct approach on Chrome (direct mode hereafter). I think the api for the CORS mode and for the direct mode should be the exact same, so this should be an abstraction.

**Comment by inf3rno** [comment]
Ok. So the conclusion that we need some kind of CORS mode, which does not affect our tests, but helps to test a server, which is maintained by us. So we have 3 modes:

 - cors mode - for different origin server
 - direct mode - for Chrome with disabled security
 - browser mode - if we start the tests from browser directly instead of using Karma

I'll rename these modes later to have more expressive names. I'll save this issue for later. Until 0.4.0 we work only on the direct mode to be able to run our tests in Chrome. Later I'll revisit this and decide about the browser mode and the implementation of the CORS mode too.

**Comment by inf3rno** [comment]
I found meanwhile that I was thinking on the eval variant of JSONP, which sends function calls from the server to the client e.g. `doSomething(arg1, arg2, ...);`. A more sophisticated solution was to send only data to the client e.g. `receive(data, data, data, ...);`. We need to run something like the first on the loaded page with a lot of sugar, but we need to make it clear that our users cannot use variables which are out of scope. Hmm phpunit had the same problem...

**Comment by inf3rno** [comment]
I think this whole thing is bigger than having a CORS mode, so I'll create a new issue for it and I close this one.
